joint_state_publisher/joint_state_publisher/joint_state_publisher.py

```diff
@@ -0,0 +1,399 @@
+                joint['position'] = position
+            if velocity is not None:
+                joint['velocity'] = velocity
+            if effort is not None:
+                joint['effort'] = effort
+
+        if self.source_update_cb is not None:
+            self.source_update_cb()
+
+    def set_source_update_cb(self, user_cb):
+        self.source_update_cb = user_cb
+
+    def timer_callback(self):
+        # Publish Joint States
+        msg = sensor_msgs.msg.JointState()
+        msg.header.stamp = self.get_clock().now().to_msg()
+
+        if self.delta > 0:
+            self.update(self.delta)
+
+        # Initialize msg.position, msg.velocity, and msg.effort.
+        has_position = len(self.dependent_joints.items()) > 0
+        has_velocity = False
+        has_effort = False
+        for name, joint in self.free_joints.items():
+            if not has_position and 'position' in joint:
+                has_position = True
+            if not has_velocity and 'velocity' in joint:
+                has_velocity = True
+            if not has_effort and 'effort' in joint:
+                has_effort = True
+        num_joints = (len(self.free_joints.items()) +
+                      len(self.dependent_joints.items()))
+        if has_position:
+            msg.position = num_joints * [0.0]
+        if has_velocity:
+            msg.velocity = num_joints * [0.0]
+        if has_effort:
+            msg.effort = num_joints * [0.0]
+
+        for i, name in enumerate(self.joint_list):
+            msg.name.append(str(name))
+            joint = None
+
+            # Add Free Joint
+            if name in self.free_joints:
+                joint = self.free_joints[name]
+                factor = 1
+                offset = 0
+            # Add Dependent Joint
+            elif name in self.dependent_joints:
+                param = self.dependent_joints[name]
+                parent = param['parent']
+                factor = param.get('factor', 1.0)
+                offset = param.get('offset', 0.0)
+                # Handle recursive mimic chain
+                recursive_mimic_chain_joints = [name]
+                while parent in self.dependent_joints:
+                    if parent in recursive_mimic_chain_joints:
+                        error_message = 'Found an infinite recursive mimic chain'
+                        self.get_logger().error('%s: [%s, %s]', error_message, ', '.join(recursive_mimic_chain_joints), parent)
+                        sys.exit(1)
+                    recursive_mimic_chain_joints.append(parent)
+                    param = self.dependent_joints[parent]
+                    parent = param['parent']
+                    offset += factor * param.get('offset', 0)
+                    factor *= param.get('factor', 1)
+                joint = self.free_joints[parent]
+
+            if has_position and 'position' in joint:
+                msg.position[i] = float(joint['position']) * factor + offset
```

Comment:
I get an `IndexError` here when I publish the same URDF twice on the `/robot_description` topic. It looks like `self.joint_list` has duplicate joint names, but `num_joints` was calculated from `free_joints` and `dependent_joints` which can't have duplicates because they're dictionaries, so `msg.position` is half the length. Maybe the variables that are set in `__init__` need to be set in `self.configure_robot` instead so they get cleared when a new description is published?

Reply:
This was absolutely a problem.  Fixing it turned out to be...complicated.  Commit 735f2a7 successfully fixes it, along with making the layout way more dynamic and useful (in my opinion).  It also simplifies the code a bit, at the expense of adding a new PyQt layout type (FlowLayout), and putting the Center/Randomize buttons at the top.  I think that is a pretty good tradeoff for a dynamic, reflowable UI, but please take a look and let me know what you think.